web3/eth.py

```diff
@@ -462,44 +459,81 @@ def estimate_gas_munger(
         mungers=[estimate_gas_munger]
     )
 
-    def filter(
-        self, filter_params: Optional[Union[str, FilterParams]] = None,
+    def filter_munger(
+        self,
+        filter_params: Optional[Union[str, FilterParams]] = None,
         filter_id: Optional[HexStr] = None
-    ) -> Filter:
+    ) -> Union[List[FilterParams], List[HexStr], List[str]]:
         if filter_id and filter_params:
             raise TypeError(
-                "Ambiguous invocation: provide either a `filter_params` or a `filter_id` argument. "
+                "Ambiguous invocation: provide either a `filter_params` or a `filter_id` argument. "  # noqa: E501
                 "Both were supplied."
             )
-        if is_string(filter_params):
-            if filter_params == "latest":
-                filter_id = self.web3.manager.request_blocking(
-                    RPC.eth_newBlockFilter, [],
-                )
-                return BlockFilter(self.web3, filter_id)
-            elif filter_params == "pending":
-                filter_id = self.web3.manager.request_blocking(
-                    RPC.eth_newPendingTransactionFilter, [],
-                )
-                return TransactionFilter(self.web3, filter_id)
+        if isinstance(filter_params, dict):
+            return [filter_params]
+        elif is_string(filter_params):
+            if filter_params == 'latest' or filter_params == 'pending':
+                return [filter_params]
             else:
                 raise ValueError(
                     "The filter API only accepts the values of `pending` or "
                     "`latest` for string based filters"
                 )
-        elif isinstance(filter_params, dict):
-            _filter_id = self.web3.manager.request_blocking(
-                RPC.eth_newFilter,
-                [filter_params],
-            )
-            return LogFilter(self.web3, _filter_id)
-        elif filter_id and not filter_params:
-            return LogFilter(self.web3, filter_id)
+        elif filter_id:
+            return [filter_id]
         else:
             raise TypeError("Must provide either filter_params as a string or "
                             "a valid filter object, or a filter_id as a string "
                             "or hex.")
 
+    filter: Method[Callable[..., Any]] = Method(
+        method_choice_depends_on_args=select_filter_method(
+            is_new_block_filter=RPC.eth_newBlockFilter,
+            is_new_pending_transaction_filter=RPC.eth_newPendingTransactionFilter,
+            is_new_filter=RPC.eth_newFilter,
+            is_existing_filter='existing_filter',
```

Comment:
@pipermerriam @marcgarreau would love some feedback on this janky handling of the existing filter case. If a user passes in a filter_id (ex. '0x0'), then we don't want to make the request to the server, we just want to return a `LogFilter` object which needs the module and the filter_id.  Currently, I'm handling it [here](https://github.com/ethereum/web3.py/pull/1778/files#diff-c473cbb4ff9e3ac6c462a6585df86227779156085f03fc5a37aad66f1e6e5a6eR47-R48), but it doesn't seem like it belongs there. Any ideas?

Reply:
🤔 Is there a not awful way to check for an existing filter prior to `select_filter_method` and bail out early before going through the Method workflow?

nitpicky: looks like the other `depends_on` methods use `if_condition` instead of `is_condition`, like you've got here.